master/pkg/schemas/expconf/log_pattern_config.go

```diff
@@ -6,84 +6,187 @@ import (
 
 	log "github.com/sirupsen/logrus"
 
+	"github.com/pkg/errors"
+
+	"github.com/determined-ai/determined/master/pkg/ptrs"
 	"github.com/determined-ai/determined/master/pkg/schemas"
-	"github.com/determined-ai/determined/master/pkg/union"
+	"github.com/determined-ai/determined/master/pkg/set"
 )
 
 // LogPoliciesConfigV0 is a list of log policies.
+//
+//go:generate ../gen.sh
 type LogPoliciesConfigV0 []LogPolicyV0
 
-// Merge implemenets the mergable interface.
+// WithDefaults implements the Defaultable pseudo-interface.
+func (b LogPoliciesConfigV0) WithDefaults() LogPoliciesConfigV0 {
+	defaultPolicies := LogPoliciesConfigV0{
+		LogPolicyV0{
+			RawName:    ptrs.Ptr(CUDAOOM),
+			RawPattern: ptrs.Ptr(CUDAOOMPattern),
+		},
+		LogPolicyV0{
+			RawName:    ptrs.Ptr(ECCError),
+			RawPattern: ptrs.Ptr(ECCErrorPattern),
+		},
+	}
+
+	return schemas.Merge(b, defaultPolicies)
+}
+
+// Merge implements the Mergable pseudo-interface.
+// Union merge log policies.
+// Special cases:
+// 1. We may see policies with different names same patterns. We keep both of them, but which name
+// will be shown in the UI is undefined.
+// 2. There could be policies with the same name different patterns. We save the one with the higher priority.
+// Unsetting default values depends on this behavior.
+// 2.1 Polices that don't have name but have same patterns is a special case. These are legacy
+// policies.
 func (b LogPoliciesConfigV0) Merge(
-	other LogPoliciesConfigV0,
+	src LogPoliciesConfigV0,
 ) LogPoliciesConfigV0 {
-	var out LogPoliciesConfigV0
-	seen := make(map[string]bool)
-	for _, p := range append(other, b...) {
-		json, err := json.Marshal(p)
-		if err != nil {
-			log.Errorf("marshaling error %+v %v", p, err)
+	if src == nil && b == nil {
+		return nil
+	}
+
+	// Keep everything in b unconditionally.
+	out := append(LogPoliciesConfigV0{}, b...)
+
+	names := set.New[string]()
+	unnamedPolicies := set.New[string]()
+	for _, lp := range b {
+		if lp.RawName == nil {
+			json, err := json.Marshal(lp)
+			if err != nil {
+				log.Errorf("marshaling error %+v %v", lp, err)
+			}
+			unnamedPolicies.Insert(string(json))
```

Comment:
Can you use the Sprintf encoding instead?  IMO you shouldn't be doing anything here that even _can_ return an error.  Not to mention I think it'll be faster and less code.

Reply:
Replace Unmarshal with Sprintf.